habitat-lab/habitat/datasets/rearrange/navmesh_utils.py

```diff
@@ -10,6 +10,137 @@
 from habitat_sim.physics import VelocityControl
 
 
+def snap_point_is_occluded(
+    target: mn.Vector3,
+    snap_point: mn.Vector3,
+    height: float,
+    sim: habitat_sim.Simulator,
+    granularity: float = 0.2,
+    target_object_id: Optional[int] = None,
+) -> bool:
+    """
+    Uses raycasting to check whether a target is occluded given a navmesh snap point.
+
+    :property target: The 3D position whic should be unoccluded from the snap point.
+    :property snap_point: The navmesh snap point under consideration.
+    :property height: The height of the agent. Given navmesh snap point is grounded, the maximum height from which a visibility check should indicate non-occlusion. First check starts from this height.
+    :property sim: The Simulator instance.
+    :property granularity: The distance between raycast samples. Finer granularity is more accurate, but more expensive.
+    :property target_object_id: An optional object id which should be ignored in occlusion check.
+
+    NOTE: If agent's eye height is known and only that height should be considered, provide eye height and granulatiry > height for fastest check.
+
+    :return: whether or not the target is considered occluded from the snap_point.
+    """
+    # start from the top, assuming the agent's eyes are not at the bottom.
+    cur_height = height
+    while cur_height > 0:
+        ray = habitat_sim.geo.Ray()
+        ray.origin = snap_point + mn.Vector3(0, cur_height, 0)
+        cur_height -= granularity
+        ray.direction = target - ray.origin
+        raycast_results = sim.cast_ray(ray)
+        # distance of 1 is the displacement between the two points
+        if (
+            raycast_results.has_hits()
+            and raycast_results.hits[0].ray_distance < 1
+            and target_object_id is not None
+        ):
+            if raycast_results.hits[0].object_id == target_object_id:
```

Comment:
Will this be a little bit too hard? Since that if the object is placed in the location that there is a cover on top of it (e.g., shelf), then the robot needs to lower the arm until it sees it? Not sure in this case, how should we treat that?

Reply:
Or something like `norm(raycast_results.hits[0].translation - target_object_translaton) <= threshold`, then we set `true`?

Reply:
In general this function checks all the heights from configured max to floor, so it gives you a single binary for whether there exists a visible height or not. 
If you want to check a specific height, then pass that one and set the granularity >= to height. You could then get a binary flag for whether the current state allow visibility.